### PR TITLE
Integrate kubevirt

### PR DESCRIPTION
**Pull request by SchSeba**
Added the ability to allocate and save mac address for kubevirt virtual machines objects.

this way all the time a VM is configure in the cluster even if it's stop the mac address will be reserve for him. (allow the virtual machine to have the same mac address after restart)

minor changes:
* update the godep file
* fix a bug with the latest available mac

---
This change is [Reviewable](https://reviewable.io/reviews/schseba/kubemacpool/7)

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -13,22 +13,20 @@ required = [
...
-
 [[constraint]]
   name="sigs.k8s.io/controller-runtime"
-  version="v0.1.1"
+  version="v0.1.10"
 
 [[constraint]]
   name="sigs.k8s.io/controller-tools"
-  version="v0.1.1"
+  version="v0.1.9"
```
Comment (phoracek) [inline Gopkg.toml]: I'm just guessing, but could that be, that controler-runtime and -tools should be set to the same version? Maybe they have some dependencies between one another?

File: Gopkg.toml [hunk 2]
```diff
@@ -39,3 +37,13 @@ version="v1.4.7"
 [[constraint]]
   name = "github.com/intel/multus-cni"
   branch = "master"
+
+[[override]]
+  name = "kubevirt.io/containerized-data-importer"
+  source = "https://github.com/kubevirt/containerized-data-importer.git"
+  version = "1.5.1"
+
+[[override]]
```
Comment (phoracek) [inline Gopkg.toml]: why override? shouldn't this be constraint? the same for CDI.

File: Gopkg.toml [hunk 3]
```diff
@@ -39,3 +37,13 @@ version="v1.4.7"
 [[constraint]]
   name = "github.com/intel/multus-cni"
   branch = "master"
+
+[[override]]
```
Comment (phoracek) [inline Gopkg.toml]: why do we need CDI?

File: pkg/controller/add_virtualmachine.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/*
+Copyright 2018 The KubeMacPool Authors.
```
Comment (phoracek) [inline pkg/controller/add_virtualmachine.go]: 2019

File: pkg/controller/virtualmachine/virtualmachine_controller.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2018 The KubeMacPool Authors.
```
Comment (phoracek) [inline pkg/controller/virtualmachine/virtualmachine_controller.go]: 2019

File: Gopkg.toml [hunk 6]
```diff
@@ -13,22 +13,20 @@ required = [
     "sigs.k8s.io/controller-runtime/pkg/runtime/signals",
     "sigs.k8s.io/controller-runtime/pkg/source",
     "sigs.k8s.io/testing_frameworks/integration", # for integration testing
-	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1",
-    ]
+    "github.com/go-openapi/spec",
+    "github.com/kubevirt/kubevirt",
```
Comment (phoracek) [inline Gopkg.toml]: You should drop this. I think that this way you make `dep` download the who kubevirt. If you drop it, it would only put the client to the vendor directory.

File: cmd/manager/main.go [hunk 7]
```diff
@@ -56,58 +57,80 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+	log.Info("kubevirt was not found in the cluster start a watching process")
+	for _ = range time.Tick(5 * time.Second) {
+		kubevirtExist := checkForKubevirt(kubeClient)
+		log.V(1).Info("kubevirt exist in the cluster", "kubevirtExist", kubevirtExist)
+		if kubevirtExist {
+			log.Info("found kubevirt restarting the manager")
+			close(reloadChannel)
+			break
+		}
+	}
```
Comment (phoracek) [inline cmd/manager/main.go]: if kubevirt is not found in 5 seconds, we switch off the support, right?

File: cmd/manager/main.go [hunk 8]
```diff
@@ -56,58 +57,80 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+}
+
+func waitforsignal() {
```
Comment (phoracek) [inline cmd/manager/main.go]: waitForSignal

File: cmd/manager/main.go [hunk 9]
```diff
@@ -56,58 +57,80 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+	startPoolRange := loadMacAddressFromEnvVar(poolmanager.StartPoolRangeEnv)
+	endPoolRange := loadMacAddressFromEnvVar(poolmanager.EndPoolRangeEnv)
+
+	isKubevirtExist := checkForKubevirt(clientset)
+	poolManager, err := poolmanager.NewPoolManager(clientset, startPoolRange, endPoolRange, isKubevirtExist)
+	ExitIfError(err, "unable to create pool manager")
+	if !isKubevirtExist {
+		go waitForKubevirt(clientset)
 	}
+	go waitforsignal()
```
Comment (phoracek) [inline cmd/manager/main.go]: could you comment this please? the logic of kubevirt signals and this is too complicated for me

File: cmd/manager/main.go [hunk 10]
```diff
@@ -56,58 +57,80 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, os.Kill)
+	<-c
+	close(reloadChannel)
+}
+
 func main() {
-	var metricsAddr string
-	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8090", "The address the metric endpoint binds to.")
```
Comment (phoracek) [inline cmd/manager/main.go]: why did you change the port?

Comment (phoracek) [inline cmd/manager/main.go]: why did you change the port?

File: cmd/manager/main.go [hunk 11]
```diff
@@ -56,58 +57,80 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+	var logType string
+	flag.StringVar(&logType, "v", "production", "Log type (debug/production).")
 	flag.Parse()
-	logf.SetLogger(logf.ZapLogger(false))
+
+	if logType == "debug" {
```
Comment (phoracek) [inline cmd/manager/main.go]: please move the logging update to a separate commit

File: cmd/manager/main.go [hunk 12]
```diff
@@ -56,58 +57,80 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+	ExitIfError(err, "unable to create a kubernetes client")
+	// setup Pool Manager
+	startPoolRange := loadMacAddressFromEnvVar(poolmanager.StartPoolRangeEnv)
+	endPoolRange := loadMacAddressFromEnvVar(poolmanager.EndPoolRangeEnv)
+
+	isKubevirtExist := checkForKubevirt(clientset)
+	poolManager, err := poolmanager.NewPoolManager(clientset, startPoolRange, endPoolRange, isKubevirtExist)
+	ExitIfError(err, "unable to create pool manager")
+	if !isKubevirtExist {
+		go waitForKubevirt(clientset)
```
Comment (phoracek) [inline cmd/manager/main.go]: you wait for kubevirt after you initialized poolManager with `isKubevirtExist`, is that intended?

File: config/manager/manager.yaml [hunk 13]
```diff
@@ -79,15 +80,15 @@ spec:
...
           name: cert
           readOnly: true
-      terminationGracePeriodSeconds: 10
+      terminationGracePeriodSeconds: 5
       volumes:
       - name: cert
         secret:
           defaultMode: 420
-          secretName: webhook-server-secret
+          secretName: webhook-secret
```
Comment (phoracek) [inline config/manager/manager.yaml]: could you move manifest changes unrelated to kubevirt to a separate commit?

File: pkg/controller/pod/pod_controller.go [hunk 14]
```diff
@@ -72,18 +72,14 @@ type ReconcilePolicy struct {
...
-	log.V(10).Info("Get Pod event in controller")
+	log.V(1).Info("got a pod event in the controller")
 	instance := &corev1.Pod{}
 	err := r.Get(context.TODO(), request.NamespacedName, instance)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
-			err := r.poolManager.ReleaseMac(fmt.Sprintf("%s-%s", request.Namespace, request.Name))
+			err := r.poolManager.ReleaseMac(fmt.Sprintf("%s/%s", request.Namespace, request.Name))
```
Comment (phoracek) [inline pkg/controller/pod/pod_controller.go]: i like the cleanup, but please move it to a separate commit

File: pkg/pool-manager/pool.go [hunk 15]
```diff
@@ -24,28 +26,32 @@ const (
...
 type PoolManager struct {
-	client.Client                       // kubernetes client
-	startRange      net.HardwareAddr    // fist mac in range
-	endRange        net.HardwareAddr    // last mac in range
-	currentMac      net.HardwareAddr    // last given mac
-	macPoolMap      map[string]bool     // allocated mac map
-	podToMacPoolMap map[string][]string // map for namespace-pod name and a list of allocated mac addresses
-	poolMutex       sync.Mutex          // mutex for allocation an release
-	isLeader        bool                // leader boolean
+	KubeClient      kubernetes.Interface // kubernetes client
```
Comment (phoracek) [inline pkg/pool-manager/pool.go]: why is this public?

File: pkg/pool-manager/pool.go [hunk 16]
```diff
@@ -126,6 +175,34 @@ func (p *PoolManager) ReleaseMac(podName string) error {
 	return nil
 }
 
+func (p *PoolManager) ReleaseVirtualMachineMac(virtualMachineName string) error {
```
Comment (phoracek) [inline pkg/pool-manager/pool.go]: would it be possible to keep only the main pool logic here and move Pod and VM specific functions/methos to their dedicated files?

File: Gopkg.toml (hunk 1 above)
Comment (SchSeba) [reply Gopkg.toml]: they have different versions for latest.

File: pkg/pool-manager/pool.go (hunk 16 above)
Comment (SchSeba) [reply pkg/pool-manager/pool.go]: done

File: Dockerfile [hunk 17]
```diff
@@ -3,9 +3,9 @@ FROM golang:1.11 as builder
 
 # Copy in the go src
 WORKDIR /go/src/github.com/SchSeba/kubemacpool
-COPY pkg/    pkg/
-COPY cmd/    cmd/
 COPY vendor/ vendor/
+COPY cmd/    cmd/
```
Comment (phoracek) [inline Dockerfile]: Could your create a commit just changing the order?

I won't bug about separate PRs, but we should keep at least commits granular. Plus you get better score on the contributors board ;)

File: Gopkg.toml (hunk 6 above)
Comment (phoracek) [reply Gopkg.toml]: I see you removed these imports in a following commit. Could you not introduce them at all?

File: Gopkg.toml (hunk 3 above)
Comment (phoracek) [reply Gopkg.toml]: If we don't need it and only kubevirt does, `make dep` should make sure that it is pulled. We however, don't need it here. I noticed you removed this dependency in a following commit. Could you not introduce it at all?

File: cluster/k8s-multus-1.13.3/provider.sh [hunk 18]
```diff
@@ -2,7 +2,7 @@
 
 set -e
 
-image="k8s-multus-1.13.3@sha256:d562b9c7d0192fdf7b2570887d05191bc7ecbf75555b5bc0ddb9782eab5a7517"
+image="k8s-multus-1.13.3@sha256:2ae2dfd20adf163e9cb3923c713932353670692d119f330e547aa1881872de58"
```
Comment (phoracek) [inline cluster/k8s-multus-1.13.3/provider.sh]: Deserves separate commit. What if it breaks our builds and the we are trying to find the commit that broke it.

File: cmd/manager/main.go (hunk 10 above)
Comment (phoracek) [reply cmd/manager/main.go]: this change seems unrelated and should be in a separate commit

**Comment by SchSeba** [comment]
Done with the commits rework 

@phoracek can you please take a look if this is better now?

File: cmd/manager/main.go [hunk 19]
```diff
@@ -23,8 +23,8 @@ import (
 	"net"
 	"os"
 
-	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
-
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/kubernetes"
```
Comment (phoracek) [inline cmd/manager/main.go]: unrelated to this change (btw you can do interactive git rebase -i to eddit just this single commit)

File: cmd/manager/main.go [hunk 20]
```diff
@@ -23,8 +23,8 @@ import (
 	"net"
 	"os"
 
-	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
-
+	"github.com/go-logr/logr"
```
Comment (phoracek) [inline cmd/manager/main.go]: i don't see logr used anywhere, just logf. on top of that, it is already imported

File: pkg/pool-manager/pool.go [hunk 21]
```diff
@@ -190,6 +190,12 @@ func (p *PoolManager) getFreeMac() (net.HardwareAddr, error) {
 			}
 		}
 
+		// Use also the last mac address in the pool
```
Comment (phoracek) [inline pkg/pool-manager/pool.go]: could you please elaborate what the problem was in the commit message? (you can also edit commit message in the interactive rebase)

File: Gopkg.toml [hunk 22]
```diff
@@ -35,3 +35,8 @@ version="v1.4.7"
 [[constraint]]
   name = "github.com/intel/multus-cni"
   branch = "master"
+
```
Comment (phoracek) [inline Gopkg.toml]: looks like you forgot to `dep ensure` again maybe? there are still CDI sources. on top of that, the constraint should be added only with code that indeed uses kubevirt.

or did you just commit it in two stages?

finally, keeping vendor/ commit separated is good for reviews when commiting vendor/ causes GitHub not to show the whole diff. however, i think that when possible, it should be added with the code that uses it. not sure if it would help while reverting, but i feel it makes more sense. won't fight for it though

File: pkg/pool-manager/pool.go [hunk 23]
```diff
@@ -216,13 +100,14 @@ func (p *PoolManager) setAsLeader() {
 }
 
 func (p *PoolManager) InitMaps() error {
-	pods := &corev1.PodList{}
-	err := p.List(context.TODO(), &client.ListOptions{}, pods)
+	log.V(1).Info("start InitMaps")
```
Comment (phoracek) [inline pkg/pool-manager/pool.go]: this does not say much when someone looks in logs, what are InitMaps?

File: pkg/controller/pod/pod_controller.go (hunk 14 above)
Comment (SchSeba) [reply pkg/controller/pod/pod_controller.go]: done

File: pkg/pool-manager/pool.go (hunk 21 above)
Comment (SchSeba) [reply pkg/pool-manager/pool.go]: Done

File: pkg/pool-manager/pool.go (hunk 23 above)
Comment (SchSeba) [reply pkg/pool-manager/pool.go]: done

File: Gopkg.toml (hunk 22 above)
Comment (SchSeba) [reply Gopkg.toml]: I remove the CDI folder from vendor and rerun dep ensure and it appears again.

File: cmd/manager/main.go (hunk 10 above)
Comment (SchSeba) [reply cmd/manager/main.go]: unrelated change I revert it thanks!

File: pkg/pool-manager/pod_pool.go [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+	if !ok {
+		return nil
+	}
+
+	networks, err := parsePodNetworkAnnotation(networkValue, pod.Namespace)
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("pod meta data", "podMetaData", (*pod).ObjectMeta)
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: please lower the priority not to spam log with object specs.

either rework info priorities or just drop V(*), use only log.Info and leave it for future PR

File: pkg/pool-manager/pod_pool.go [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+	pod.Annotations[networksAnnotation] = string(networkListJson)
+
+	return nil
+}
+
+func (p *PoolManager) ReleaseMac(podName string) error {
+	p.poolMutex.Lock()
+	defer p.poolMutex.Unlock()
+
+	p.setAsLeader()
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: why do you set leader in method calls? shouldn't that happen during the controller initialization?

File: pkg/pool-manager/pod_pool.go [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+	"encoding/json"
+	"fmt"
+	"net"
+	"strings"
+
+	multus "github.com/intel/multus-cni/types"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func (p *PoolManager) AllocateMac(pod *corev1.Pod) error {
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: please namespace method names in this module to "pod", e.g. PodAllocateMac

File: pkg/pool-manager/pod_pool.go [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+func (p *PoolManager) ReleaseMac(podName string) error {
+	p.poolMutex.Lock()
+	defer p.poolMutex.Unlock()
+
+	p.setAsLeader()
+
+	macList, ok := p.podToMacPoolMap[podName]
+
+	if !ok {
+		log.Error(fmt.Errorf("not found"), "pod not found in the map",
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: why do you call fmt.Errorf on the message, why? does it matter if it is only for logs?

File: pkg/pool-manager/pod_pool.go [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+		return nil
+	}
+
+	for _, macAddr := range macList {
+		delete(p.macPoolMap, macAddr)
+		log.Info("released mac from pod", "mac", macAddr, "pod", podName)
+	}
+
+	delete(p.podToMacPoolMap, podName)
+	log.Info("removed pod from podToMacPoolMap", "pod", podName)
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: in other Info logs, you specify the priority. could you specify priority everywhere?

File: cmd/manager/main.go [hunk 29]
```diff
@@ -56,8 +57,39 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+		if kubevirtExist {
+			log.Info("found kubevirt restarting the manager")
+			close(reloadChannel)
+			break
+		}
+	}
+}
+
+// wait for the any interrupt to stop the manager and clean the webhook.
+func waitForSignal() {
```
Comment (phoracek) [inline cmd/manager/main.go]: the kubevirt logic is complicated as it is, could you please move this into a separate commit?

File: cmd/manager/main.go [hunk 30]
```diff
@@ -72,48 +104,35 @@ func main() {
 	// Get a config to talk to the apiserver
 	log.Info("Setting up client for manager")
 	cfg, err := config.GetConfig()
-	if err != nil {
-		log.Error(err, "unable to set up client config")
-		os.Exit(1)
+	ExitIfError(err, "unable to set up client config")
```
Comment (phoracek) [inline cmd/manager/main.go]: maybe I'm too nitpicking here, but could you please move ExitIfError to a separate commit?

i want to have kubevirt support as isolated and modular as possible, so I'd like to see only related changes to figure out whether it could be done better

File: pkg/pool-manager/pod_pool.go [hunk 31]
```diff
@@ -44,6 +44,16 @@ func (p *PoolManager) AllocateMac(pod *corev1.Pod) error {
 
 	log.V(1).Info("pod meta data", "podMetaData", (*pod).ObjectMeta)
 
+	if len(networks) == 0 {
+		return nil
+	}
+
+	if IsRelatedToKubevirt(pod) {
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: this made me think - can we change the logic so we keep only single POD macpool, we have only pod controller. Here we check whether the owner is VM and if it is, we edit the Pod definition the same way we do it with other pods.

With that we would have 1 controller. Just a bit of a special handling (if KubeVirt is the owner, we read desired mac from VMI, not from multus). Maybe I'm missing something though.

I feel bad for asking to change such a big chunk of code, but could you consider it? It would simplify the manager from "supporting 2 kinds of objects, with full blown controllers ..." to "we can read mac request from VMI too".

File: cmd/manager/main.go (hunk 20 above)
Comment (SchSeba) [reply cmd/manager/main.go]: ops you are right :)

File: pkg/pool-manager/pool.go (hunk 21 above)
Comment (SchSeba) [reply pkg/pool-manager/pool.go]: I don't want to run this if on every iteration of the loop this is why I move it outside the for loop

File: pkg/controller/pod/pod_controller.go [hunk 32]
```diff
@@ -72,11 +72,9 @@ type ReconcilePolicy struct {
 	poolManager *pool_manager.PoolManager
 }
 
-// Reconcile reads that state of the cluster for a Pod object and makes changes based on the state read
-// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+// Reconcile reads that state of the cluster for a Pod object and makes changes based on the state
 func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	// Fetch the Policy instance
-	log.V(10).Info("Get Pod event in controller")
+	log.V(1).Info("got a pod event in the controller")
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/controller/pod/pod_controller.go]: the new version of logs have only two configuration 
1. production (log.Info)
2. debug (log.V(1).Info)

the log commands that have V(1) will be printed only on debug mode.

File: pkg/pool-manager/pod_pool.go (hunk 24 above)
Comment (SchSeba) [reply pkg/pool-manager/pod_pool.go]: the new version of logs have only two configuration

production (log.Info)
debug (log.V(1).Info)
the log commands that have V(1) will be printed only on debug mode.

File: pkg/pool-manager/pod_pool.go (hunk 25 above)
Comment (SchSeba) [reply pkg/pool-manager/pod_pool.go]: set leader happens when we get the first call this tell as the now the running process is the leader

File: pkg/pool-manager/pod_pool.go (hunk 27 above)
Comment (SchSeba) [reply pkg/pool-manager/pod_pool.go]: this is how error works in the log package.
```golang
Error(err error, msg string, keysAndValues ...interface{})
```

File: pkg/pool-manager/pod_pool.go (hunk 28 above)
Comment (SchSeba) [reply pkg/pool-manager/pod_pool.go]: the new version of logs have only two configuration

production (log.Info)
debug (log.V(1).Info)
the log commands that have V(1) will be printed only on debug mode.

File: pkg/pool-manager/pod_pool.go (hunk 31 above)
Comment (SchSeba) [reply pkg/pool-manager/pod_pool.go]: as we spoke offline we can't do it because we need to allocate on the virtual machine webhook and not wait for a pod to be created.

File: cmd/manager/main.go (hunk 30 above)
Comment (SchSeba) [reply cmd/manager/main.go]: done new commit

File: cmd/manager/main.go (hunk 29 above)
Comment (SchSeba) [reply cmd/manager/main.go]: done

File: pkg/pool-manager/pool.go (hunk 21 above)
Comment (phoracek) [reply pkg/pool-manager/pool.go]: But there is no change compared to the orignal code IIUIC. The code I suggested just transforms your while loop to **do** while. In the current code you just check `currentMac.String() != p.endRange.String()` before handling the address, i suggest to move it to the end. Maybe I'm missing something

File: pkg/pool-manager/pod_pool.go (hunk 26 above)
Comment (phoracek) [reply pkg/pool-manager/pod_pool.go]: @SchSeba ^^

File: cmd/manager/main.go (hunk 19 above)
Comment (phoracek) [reply cmd/manager/main.go]: @SchSeba ^^

File: cmd/manager/main.go (hunk 10 above)
Comment (phoracek) [reply cmd/manager/main.go]: You reverted it in a different commit than you introduced it, please revert it here.

File: cmd/manager/main.go [hunk 33]
```diff
@@ -89,8 +112,12 @@ func main() {
 	startPoolRange := loadMacAddressFromEnvVar(poolmanager.StartPoolRangeEnv)
 	endPoolRange := loadMacAddressFromEnvVar(poolmanager.EndPoolRangeEnv)
 
-	poolManager, err := poolmanager.NewPoolManager(clientset, startPoolRange, endPoolRange)
+	isKubevirtExist := checkForKubevirt(clientset)
```
Comment (phoracek) [inline cmd/manager/main.go]: isKubevirtInstalled

File: pkg/pool-manager/virtualmachine_pool.go [hunk 34]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	log.V(1).Info("virtual machine data",
+		"name", virtualMachine.Name,
+		"namespace", virtualMachine.Namespace,
+		"interfaces", virtualMachine.Spec.Template.Spec.Domain.Devices.Interfaces)
+
+	for idx, iface := range virtualMachine.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if iface.Masquerade == nil && iface.Slirp == nil && networks[iface.Name].Multus == nil {
+			log.Info("for k8s network mac address will be set only for masquerade and slirp")
```
Comment (phoracek) [inline pkg/pool-manager/virtualmachine_pool.go]: could you rephrase it? i don't understand the message

File: pkg/webhook/pod/pod.go [hunk 35]
```diff
@@ -99,8 +85,9 @@ func (a *podAnnotator) mutatePodsFn(ctx context.Context, pod *corev1.Pod) error
 	if pod.Annotations == nil {
 		return nil
 	}
-	log.V(10).Info("Get mutate pod event")
-	return a.poolManager.AllocateMac(pod)
+
+	log.Info("got a mutate pod event", "podName", pod.Name, "podNamespace", pod.Namespace)
```
Comment (phoracek) [inline pkg/webhook/pod/pod.go]: should be debug level

File: cmd/manager/main.go [hunk 36]
```diff
@@ -56,58 +57,82 @@ func loadMacAddressFromEnvVar(envName string) net.HardwareAddr {
...
+	if result.Error() == nil {
+		return true
+	}
+
+	return false
+}
+
+// Check for Kubevirt CRD to be available
+func waitForKubevirt(kubeClient *kubernetes.Clientset) {
+	log.Info("kubevirt was not found in the cluster start a watching process")
```
Comment (phoracek) [inline cmd/manager/main.go]: this log should be on caller's level, not in this function

File: pkg/pool-manager/pod_pool.go [hunk 37]
```diff
@@ -0,0 +1,179 @@
...
+
+	log.V(1).Info("pod meta data", "podMetaData", (*pod).ObjectMeta)
+
+	if len(networks) == 0 {
+		return nil
+	}
+
+	if IsRelatedToKubevirt(pod) {
+		// nothing to do here. the mac is already by allocated by the virtual machine webhook
+		log.Info("This pod have ownerReferences from kubevirt skipping")
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: should be debug level

File: pkg/pool-manager/pod_pool.go [hunk 38]
```diff
@@ -0,0 +1,179 @@
...
+		return nil
+	}
+
+	for _, macAddr := range macList {
+		delete(p.macPoolMap, macAddr)
+		log.Info("released mac from pod", "mac", macAddr, "pod", podName)
+	}
+
+	delete(p.podToMacPoolMap, podName)
+	log.Info("removed pod from podToMacPoolMap", "pod", podName)
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: should be debug level, this is internal logic and has no value to user

File: pkg/pool-manager/pod_pool.go [hunk 39]
```diff
@@ -0,0 +1,179 @@
...
+
+		return err
+	}
+
+	p.macPoolMap[requestedMac] = true
+	if p.podToMacPoolMap[podNamespaced(pod)] == nil {
+		p.podToMacPoolMap[podNamespaced(pod)] = []string{}
+	}
+	p.podToMacPoolMap[podNamespaced(pod)] = append(p.podToMacPoolMap[podNamespaced(pod)], requestedMac)
+	log.V(1).Info("requested mac was allocated for pod",
```
Comment (phoracek) [inline pkg/pool-manager/pod_pool.go]: log level should be consistent with https://github.com/SchSeba/kubemacpool/pull/7/files#diff-f178743f06898ca1ae8f344f45065660R104 either make both debug or production level

File: pkg/pool-manager/pool.go [hunk 40]
```diff
@@ -24,154 +36,56 @@ const (
...
+	// first iteration from current mac to last mac in the range
+	// second iteration from first mac in the range to the latest one
 	for idx := 0; idx <= 1; idx++ {
-		for ; currentMac.String() != p.endRange.String(); currentMac = getNextMac(currentMac) {
+
+		// This loop runs from the current mac to the last one in the range
+		for {
+			currentMac := getNextMac(currentMac)
 			if _, ok := p.macPoolMap[currentMac.String()]; !ok {
 				log.Info("found unused mac", "mac", currentMac)
```
Comment (phoracek) [inline pkg/pool-manager/pool.go]: debug level

File: cmd/manager/main.go (hunk 36 above)
Comment (SchSeba) [reply cmd/manager/main.go]: done

File: pkg/pool-manager/pod_pool.go (hunk 37 above)
Comment (SchSeba) [reply pkg/pool-manager/pod_pool.go]: done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 41]
```diff
@@ -0,0 +1,162 @@
...
+	kubevirt "kubevirt.io/kubevirt/pkg/api/v1"
+)
+
+func (p *PoolManager) AllocateVirtualMachineMac(virtualMachine *kubevirt.VirtualMachine) error {
+	p.poolMutex.Lock()
+	defer p.poolMutex.Unlock()
+
+	p.setAsLeader()
+	if len(virtualMachine.Spec.Template.Spec.Domain.Devices.Interfaces) == 0 {
+		log.V(1).Info("failed to find interfaces for virtual machine", "virtualMachine", virtualMachine)
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/pool-manager/virtualmachine_pool.go]: done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 42]
```diff
@@ -0,0 +1,162 @@
...
+		return nil
+	}
+
+	if len(virtualMachine.Spec.Template.Spec.Networks) == 0 {
+		log.Info("no networks found for virtual machine skipping mac allocation", "name", virtualMachine.Name,
+			"namespace", virtualMachine.Namespace)
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/pool-manager/virtualmachine_pool.go]: right!

File: pkg/pool-manager/virtualmachine_pool.go (hunk 34 above)
Comment (SchSeba) [reply pkg/pool-manager/virtualmachine_pool.go]: done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 43]
```diff
@@ -0,0 +1,162 @@
...
+	p.vmToMacPoolMap[vmNamespaced(virtualMachine)] = append(p.vmToMacPoolMap[vmNamespaced(virtualMachine)], requestedMac)
+	log.V(1).Info("requested mac was allocated for virtual machine",
+		"requestedMap", requestedMac,
+		"virtualMachineName", virtualMachine.Name,
+		"virtualMachineNamespace", virtualMachine.Namespace)
+
+	return nil
+}
+
+func (p *PoolManager) IsKubevirtEnable() bool {
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/pool-manager/virtualmachine_pool.go]: done

File: pkg/pool-manager/virtualmachine_pool.go [hunk 44]
```diff
@@ -0,0 +1,162 @@
...
+	return p.isKubevirt
+}
+
+func IsRelatedToKubevirt(pod *corev1.Pod) bool {
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/pool-manager/virtualmachine_pool.go]: done

**Comment by phoracek** [comment]
@SchSeba there are still a few comments you haven't respond to. Could you please take a look?

**Comment by phoracek** [comment]
LGTM!

**Comment by SchSeba** [comment]
@phoracek Thanks for all the help!
